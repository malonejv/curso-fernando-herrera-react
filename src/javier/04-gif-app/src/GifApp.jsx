import React, { useState } from 'react'
import { AddCategory } from './components/AddCategory';
import { GifGrid } from './components/GifGrid';

export const GifApp = () => {

    const [categories, setCategories] = useState([])

    const onNewCategory = (newCategory) => {
        if (!categories.includes(newCategory))
            setCategories([newCategory, ...categories]);
    }

    return (
        <>
            {/* TÃ­tulo */}
            <h1>GifApp</h1>

            {/* Input */}
            <AddCategory
                // onPressEnter={setCategories} 
                onNewCategory={onNewCategory}
            />

            {/* Listado de Gif */}
            <GifGrid categories={categories} />

        </>
    )
}
