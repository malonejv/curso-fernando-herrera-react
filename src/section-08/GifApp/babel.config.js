// babel.config.js
/*
Explicación:

presets (fuera de env): Aquí pones los presets que son comunes a todos los entornos (en tu caso, @babel/preset-react).
plugins (fuera de env): Lo mismo para los plugins comunes.
env object: 
  Contiene objetos con nombres de entorno (ej. development, production, test). 
  Cuando Babel se ejecuta, busca la variable de entorno BABEL_ENV. Si no está definida, busca NODE_ENV. Si tampoco está, por defecto es development.
  Cuando BABEL_ENV o NODE_ENV coincidan con una de las claves dentro de env, Babel fusionará esa configuración con la configuración base.

*/
module.exports = {
  presets: [
    ['@babel/preset-react', { runtime: 'automatic' }]
  ],
  plugins: [
    // Plugins comunes para desarrollo y producción
  ],
  env: {
    development: {
      // targets: { esmodules: true } podría ir aquí si usas webpack-dev-server
    },
    production: {
      presets: [
        ['@babel/preset-env', {
          // Targets para navegadores, quizás desde .browserslistrc
          // Si no tienes un .browserslistrc, puedes definir aquí:
          // targets: "> 0.25%, not dead", // Por ejemplo
          // o especificar versiones de navegadores concretas
          // targets: { chrome: '80', firefox: '78' }
        }]
      ],
      plugins: [
        // Plugins para optimización de producción (ej. tree shaking)
      ]
    },
    test: {
      presets: [
        ['@babel/preset-env', {
          targets: { node: 'current' }
        }]
      ],
      plugins: [
        // Plugins específicos para tests (ej. babel-plugin-transform-require-context)
      ]
    }
  }
};