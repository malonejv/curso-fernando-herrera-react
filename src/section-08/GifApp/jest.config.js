module.exports = {
  roots: ['<rootDir>/src'],
  projects: [
    {
      displayName: 'unit',
      testMatch: ['**/unit-tests/**/*.test.{js,jsx}'],
      setupFiles: ['./jest.setup.js', './jest.setup.unit-test.js'],
      setupFilesAfterEnv: ['dotenv/config'],
      testEnvironment: 'jsdom',
      moduleNameMapper: {
        '^config/env$': '<rootDir>/src/common/config/env.testing.js',
        '^config/(.*)$': '<rootDir>/src/common/config/$1',
        "\\.(css|less|scss|sass)$": "identity-obj-proxy"
      },
    },
    {
      displayName: 'integration',
      testMatch: ['**/integration-tests/**/*.test.{js,jsx}'],
      setupFiles: ['./jest.setup.js'],
      setupFilesAfterEnv: ['dotenv/config'],
      testEnvironment: 'jsdom',
      moduleNameMapper: {
        '^config/env$': '<rootDir>/src/common/config/env.testing.js',
        '^config/(.*)$': '<rootDir>/src/common/config/$1',
        "\\.(css|less|scss|sass)$": "identity-obj-proxy"
      },
    },
  ],
  transform: {
    "^.+\\.(js|jsx)$": "babel-jest"
  },
  extensionsToTreatAsEsm: ['.jsx'],
  moduleFileExtensions: ['js', 'jsx'],
};