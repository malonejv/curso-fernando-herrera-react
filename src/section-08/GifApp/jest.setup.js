import 'whatwg-fetch';

jest.spyOn(console, "error").mockImplementation(() => {});