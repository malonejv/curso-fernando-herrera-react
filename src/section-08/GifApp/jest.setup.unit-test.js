import { jest } from '@jest/globals';

jest.mock('./src/common/config/env', () => ({
  apiKey: 'test-api-key'
}));