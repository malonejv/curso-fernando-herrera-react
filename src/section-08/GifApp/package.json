{
  "name": "gifapp",
  "private": true,
  "version": "0.0.0",
  "scripts": {
    "dev": "vite",
    "build": "vite build",
    "lint": "eslint .",
    "preview": "vite preview",
    "test": "node --experimental-vm-modules ./node_modules/jest/bin/jest.js --watchAll=true --coverage --passWithNoTests"
  },
  "dependencies": {
    "prop-types": "^15.8.1",
    "react": "^19.1.0",
    "react-dom": "^19.1.0"
  },
  "devDependencies": {
    "@babel/core": "^7.27.4",
    "@babel/preset-env": "^7.27.2",
    "@babel/preset-react": "^7.27.1",
    "@eslint/js": "^9.25.0",
    "@testing-library/dom": "^10.4.0",
    "@testing-library/react": "^16.3.0",
    "@types/jest": "^29.5.14",
    "@types/react": "^19.1.2",
    "@types/react-dom": "^19.1.2",
    "@vitejs/plugin-react": "^4.4.1",
    "babel-jest": "^30.0.0-beta.3",
    "dotenv": "^16.5.0",
    "eslint": "^9.25.0",
    "eslint-plugin-react-hooks": "^5.2.0",
    "eslint-plugin-react-refresh": "^0.4.19",
    "globals": "^16.0.0",
    "identity-obj-proxy": "^3.0.0",
    "jest": "^29.7.0",
    "jest-environment-jsdom": "^30.0.0-beta.3",
    "vite": "^6.3.5",
    "whatwg-fetch": "^3.6.20"
  }
}
