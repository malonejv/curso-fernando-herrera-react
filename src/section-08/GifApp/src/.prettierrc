{
  "printWidth": 240,
  "jsxSingleQuote": false,
  "jsxBracketSameLine": true,
  "htmlWhitespaceSensitivity": "ignore",
  "trailingComma": "none",
  "useTabs": false,
  "tabWidth": 2,
  "semi": true,
  "embeddedLanguageFormatting": "auto"
}