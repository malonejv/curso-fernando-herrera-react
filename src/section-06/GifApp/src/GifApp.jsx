import { useState } from "react";
import "./GifApp.css";
import { AddCategory, GifGrid  } from "./components";
import { Assertion } from "./common/Assertion";

function GifApp() {
  const [categories, setCategories] = useState(["Random"]);

  const onNewCategory = (newCategory) => {
    try {
      Assertion.This(newCategory).IsNotNullOrEmpty("Debe ingresar una categorÃ­a.");

      try {
        Assertion.This(categories).DoesNotContain(newCategory);

        setCategories([newCategory, ...categories]);
      } catch {
        return;
      }
    } catch (error) {
      console.error(error.message);
      return;
    }
  };

  const handleRemoveCategory = (category) => {
    const updatedCategories = categories.filter((cat) => cat !== category);
    setCategories(updatedCategories);
  };
  
  return (
    <>
      <h1>GifApp</h1>
      <AddCategory onNewCategory={onNewCategory} />
      {categories.map((category) => (
        <GifGrid key={category} category={category} onRemove={handleRemoveCategory} />
      ))}
    </>
  );
}

export default GifApp;
