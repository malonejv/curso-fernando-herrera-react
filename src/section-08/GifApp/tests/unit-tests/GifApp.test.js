import React from 'react';
import { render, screen, fireEvent } from "@testing-library/react";
import GifApp from '../../src/GifApp'


describe("GifApp", () => {

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test("Should render the title", () => {
        const { container } = render(<GifApp />);
        expect(container).toMatchSnapshot();
    });

    test("Should add a new category", () => {
        render(<GifApp />);

        const input = screen.getByPlaceholderText("Agrega una categoría");
        const form = screen.getByRole("form");

        fireEvent.change(input, { target: { value: "Capibara" } });
        fireEvent.submit(form);

        // Espera que el nuevo header esté en el documento
        expect(screen.getByText("Capibara")).toBeTruthy();
    });

    test("Should not add empty category", () => {
        render(<GifApp />);
        const form = screen.getByRole("form");
        fireEvent.submit(form);
        expect(screen.queryByText("Debe ingresar una categoría.")).toBeTruthy();
    });

    test("Should not add duplicate category", () => {
        render(<GifApp />);
        const input = screen.getByPlaceholderText("Agrega una categoría");
        const form = screen.getByRole("form");

        fireEvent.change(input, { target: { value: "Random" } });
        fireEvent.submit(form);

        // No debe agregar otra categoría "Random"
        const headers = screen.getAllByText("Random");
        expect(headers.length).toBe(1);
    });

    test("Should remove a category", () => {
        jest.spyOn(React, 'useState').mockImplementationOnce(() => [["Random"], jest.fn()]);
        
        render(<GifApp />);
        
        const input = screen.getByPlaceholderText("Agrega una categoría");
        const form = screen.getByRole("form");
        fireEvent.change(input, { target: { value: "Capibara" } });
        fireEvent.submit(form);

        // Busca el botón de eliminar (X) del header de Capibara
        let removeButtons = screen.getAllByText("X");
        fireEvent.click(removeButtons[0]);
        

        // Ya no debe estar Capibara en el documento
        expect(screen.queryByText("Capibara")).not.toBeTruthy();
    });

});