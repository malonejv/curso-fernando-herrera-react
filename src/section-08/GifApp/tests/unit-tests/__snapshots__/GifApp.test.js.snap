// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifApp Should render the title 1`] = `
<div>
  <h1>
    GifApp
  </h1>
  <form
    aria-label="form"
  >
    <div
      class="add-category"
    >
      <div
        class="add-category-input"
      >
        <input
          class="mr-1"
          id="category"
          name="category"
          placeholder="Agrega una categorÃ­a"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary"
          id="add-button"
          type="submit"
        >
          Agregar
        </button>
      </div>
      <span
        class="error"
        id="category-error"
      />
    </div>
  </form>
  <div
    class="gif-grid"
    data-category="Random"
  >
    <h2
      class="grid-header"
    >
      Random
       
      <button>
        X
      </button>
    </h2>
    <div
      class="loading"
    >
      Cargando...
    </div>
    <div
      class="card-grid"
    />
  </div>
</div>
`;
